docs/adr/0005-multitenancy-approach-for-marain.md

```diff
@@ -0,0 +1,263 @@
+
+```
+Root tenant
+ +
+ |
+ +-> Contoso
+ |
+ +-> Litware
+ |
+ +-> WORKFLOW
+ |
+ +-> OPERATIONS
+```
+
+Contoso is licenced to use Workflow, and Litware is licenced to use both Workflow and Operations. This means that:
+- The Contoso tenant will contain storage configuration for the Workflow service (as with all this configuration, the onboarding process will default this to standard Marain storage, where data is siloed by tenant in shared storage accounts - e.g. a single Cosmos database containing a collection per tenant. However, clients can supply their own storage configuration where required).
+- The Litware tenant will contain storage configuration for both Workflow and Operations services, because it uses both directly.
+
+In addition, because both clients are licenced for workflow, they will each have a sub-tenant of the Workflow Service Tenant, containing the storage configuration that should be used with the Operations service. The Operations service does not have any sub-tenants because it does not have dependencies on any other Marain services:
+
+```
+Root tenant
+ +
+ |
+ +-> Contoso
+ |     +-> (Workflow storage configuration)
+ |     +-> (The Id of the WORKFLOW+Contoso sub-tenant for the Workflow service to use)
+ |
+ +-> Litware
+ |     +-> (Workflow storage configuration)
+ |     +-> (The Id of the WORKFLOW+Litware sub-tenant for the Workflow service to use)
+ |     +-> (Operations storage configuration)
+ |
+ +-> WORKFLOW
+ |     |
+ |     +-> WORKFLOW+Contoso
+ |     |     +-> (Operations storage configuration)
+ |     |
+ |     +-> WORKFLOW+Litware
+ |           +-> (Operations storage configuration)
+ |
+ +-> OPERATIONS
+```
+
+As can be seen from the above, each tenant holds appropriate configuration for the services they use directly. In the case of the Client Tenants, they also hold the Id of the sub-tenant that the Workflow service will use when calling out to the Operations service on their behalf; this is necessary to avoid a costly search for the correct sub-tenant to use.
+
+You will notice from the above that Litware ends up with two sets of configuration for Operations storage; that which is employed when using the Operations service directly, and that used when calling the Workflow service and thus using the Operations service indirectly. This gives clients the maximum flexibility in controlling where their data is stored.
+
+Now let's look at a slightly more complex example. Imagine in the scenario above, there is a third service, which we'll just call the FooBar service, and that both the Workflow and Operations service are dependent on it. In addition, Contoso are licenced to use it directly. This is what the dependency graph now looks like this:
+
+```
+                          +------------+
+                          |            |
+                  +-------> Workflow   +------+-----------------+
```

Comment:
Consider using SCREAMING CAPS for the services here so that `WORKFLOW` aligns with the uppercase tenant name. It would make it easier to see which boxes here are Marain services and which are clients.